### DISCUSSION TITLE
How to write tests for axum-login?

### DISCUSSION THREAD

**Discussion by thor314** [maxcountryman/axum-login#205] [Q&A]
Integrating axum-login into a server, it's not immediately obvious what interface axum-login provides to my front-end, or how to test a login cycle. Is there any documentation here? I found this old discussion, but i'd appreciate if anyone could point me toward a more complete source of documentation on testing with axum-login, and what shape my frontend api caller should expect.
ref: https://github.com/maxcountryman/axum-login/discussions/168

**Comment by maxcountryman** [comment]
> what shape my frontend api caller should expect

I'm not exactly sure what you mean by this.

Taking a step back, axum-login doesn't have an opinion about your front end (or if you even have one): you implement an arbitrary authn (and optionally authz) workflow via the provided traits. The workflow and its details are entirely up to you.

So in that regard, beyond the discussion you linked, it's not clear there's much more to say about this.

Specific workflows may want to provide specific testing (of that logic) but that isn't really related to axum-login directly.

**Reply by thor314** [reply]
it's taken me several days to figure out how to use the interfaces of axum-login, and will likely take me the rest of today to integrate and test them. 

I find the front documentation page on the crate to be an alright start, but I'm frustrated with the state of documentation for this crate. Let me know if you would be amenable to some suggestions on how to improve them, I wouldn't want to waste your time otherwise.

**Reply by maxcountryman** [reply]
I'm happy to review PRs that add to and improve the documentation.

**Reply by thor314** [reply]
I've run into a couple weird issues, like axum-login trying to delete tokens from my database in ways that I'm uncertain how to resolve. Could I induce you to write tests for your examples? I'm uncertain I can actually resolve these on my own.
e.g., an error I've run into: 
`2024-03-26T01:18:25.840345Z DEBUG call:call:login{user.id="alice"}: sqlx::query: summary="delete from \"tower_sessions\".\"session\" where …" db.statement="\n\ndelete from\n  \"tower_sessions\".\"session\"\nwhere\n  id = $1\n" rows_affected=0 rows_returned=0 elapsed=347.467µs`
`b"{\"code\":400,\"error\":\"AuthenticationError: error returned from database: relation \\\"tower_sessions.session\\\" does not exist\"}",`

**Reply by thor314** [reply]
Or rather, I'll do it tomorrow and see if I can reproduce the issue.

**Reply by maxcountryman** [reply]
There are [integration tests](https://github.com/maxcountryman/axum-login/blob/e9597526afbe6325826b24c04af645f1e52eeae6/axum-login/tests/integration-test.rs#L53) for two of the examples; if you track down the issue and it's not covered by these then it would be great to extend the coverage.

Besides the integration tests, the crate itself is fairly extensively tested with unit-style testing.

**Reply by thor314** [reply]
I wrote a test for sqlite, threw it in pr #208, and would appreciate if you could answer a few questions.

**Reply by thor314** [reply]
I think I've found the source of my issue, documented in PR #210, mentioned 4 comments above this one.
